testtools/_deferredmatchers.py

```diff
@@ -0,0 +1,188 @@
+            on_failure=self._got_failure,
+            on_no_result=self._got_no_result,
+        )
+
+
+# XXX: The Twisted name is successResultOf. Do we want to use that name?
+def successful(matcher):
+    """Match a Deferred that has fired successfully.
+
+    For example::
+
+        fires_with_the_answer = successful(Equals(42))
+        deferred = defer.succeed(42)
+        assert_that(deferred, fires_with_the_answer)
+
+    This assertion will pass. However, if ``deferred`` had fired with a
+    different value, or had failed, or had not fired at all, then it would
+    fail.
+
+    Use this instead of
+    :py:meth:`twisted.trial.unittest.SynchronousTestCase.successResultOf`.
+
+    :param matcher: A matcher to match against the result of a
+        :class:`~twisted.internet.defer.Deferred`.
+    :return: A matcher that can be applied to a synchronous
+        :class:`~twisted.internet.defer.Deferred`.
+    """
+    return _Successful(matcher)
+
+
+class _Failed(object):
+    """Matches a Deferred that has failed."""
+
+    def __init__(self, matcher):
+        self._matcher = matcher
+
+    def _got_failure(self, deferred, failure):
+        # We have handled the failure, so suppress its output.
+        deferred.addErrback(lambda _: None)
+        return self._matcher.match(failure)
+
+    @staticmethod
+    def _got_success(deferred, success):
+        return Mismatch(
+            _u('Failure result expected on %r, found success '
+               'result (%r) instead' % (deferred, success)))
+
+    @staticmethod
+    def _got_no_result(deferred):
+        return Mismatch(
+            _u('Failure result expected on %r, found no result instead'
+               % (deferred,)))
+
+    def match(self, deferred):
+        return on_deferred_result(
+            deferred,
+            on_success=self._got_success,
+            on_failure=self._got_failure,
+            on_no_result=self._got_no_result,
+        )
+
+
+# XXX: The Twisted name is failureResultOf. Do we want to use that name?
+#
+# XXX: failureResultOf also takes an *args of expected exception types. Do we
+# want to provide that?
+def failed(matcher):
+    """Match a Deferred that has failed.
+
+    For example::
+
+        error = RuntimeError('foo')
+        fails_at_runtime = failed(Equals(error))
+        deferred = defer.fail(error)
+        assert_that(deferred, fails_at_runtime)
+
+    This assertion will pass. However, if ``deferred`` had fired successfully,
+    had failed with a different error, or had not fired at all, then it would
+    fail.
+
+    Use this instead of
+    :py:meth:`twisted.trial.unittest.SynchronousTestCase.failureResultOf`.
+
+    :param matcher: A matcher to match against the result of a failing
+        :class:`~twisted.internet.defer.Deferred`.
+    :return: A matcher that can be applied to a synchronous
+        :class:`~twisted.internet.defer.Deferred`.
+    """
+    return _Failed(matcher)
```

Comment:
To prevent subclassing so that we're not supporting subclassing as an API in the future. We could do this from documentation, but this does better on the Rusty scale.